Implemented offscreen API information into migration guides (#4765)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/offscreen/index.md

api: offscreen
---

## Manifest
You must declare the "offscreen" permission in the [extension manifest][1] to use the input.ime API. For example:

```json
{
  "name": "My extension",
  ...
  "permissions": [
    "offscreen"
  ],
  ...
}
```

> Offscreen documents are treated as web pages, and do not share the same permissions as service workers. Offscreen Windows cannot be focused, and cannot have their opener set using the [chrome.windows API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener`.

Reviewer: > Offscreen documents are treated as web pages

This isn't quite right. Offscreen documents are notably different from both typical web page and extension pages. Extensions can't load arbitrary page as an offscreen document - they have to be bundled in the extension's package. But they're also different from typical pages packed in an extension because they have very limited access to extensions APIs. 

> and do not share the same permissions as service workers

They have the same permissions. For example, the "clipboardWrite" permission applies to offscreen documents, meaning extensions can call `document.execCommand('copy')` in an offscreen document.

This suggestion is very much that: a suggestion. Feel free to use it in part, whole, or completely rewrite it.

```suggestion
Pages loaded as offscreen documents are handled differently from other types of extensions pages. The extension's permissions carry over to offscreen documents, but extension API access is heavily limited. Currently, an offscreen document can only use the `chrome.runtime` APIs to send and receive messages; all other extension APIs are not exposed. Other notable differences between offscreen documents and normal pages are as follows:

* Offscreen document's URL must be a static HTML file bundled with the extension.
* Offscreen documents cannot be focused.
* Offscreen documents cannot have their `opener` property set using the [`chrome.windows` API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener`.
```

## Examples
The following methods create and close an offscreen document. Only a single Document can be open at a time. 

```js

>      chrome.offscreen.createDocument(
>             {
>               url: chrome.runtime.getURL('off_screen.html'),
>               reasons: ['CLIPBOARD'],
>               justification: 'ignored',
>             });
> 

Reviewer: Normalize formatting.

```suggestion
chrome.offscreen.createDocument({
  url: chrome.runtime.getURL('off_screen.html'),
  reasons: ['CLIPBOARD'],
  justification: 'ignored',
});

```

api: offscreen
---

## Manifest
You must declare the "offscreen" permission in the [extension manifest][1] to use the Offscreen API. For example:

```json
{
  "name": "My extension",
  ...
  "permissions": [
    "offscreen"
  ],
  ...
}
```
Pages loaded as offscreen documents are handled differently from other types of extensions pages. The extension's permissions carry over to offscreen documents, but extension API access is heavily limited. Currently, an offscreen document can only use the `chrome.runtime` APIs to send and receive messages; all other extension APIs are not exposed. Other notable differences between offscreen documents and normal pages are as follows:

* Offscreen document's URL must be a static HTML file bundled with the extension.
* Offscreen documents cannot be focused.
* Offscreen documents cannot have their `opener` property set using the [`chrome.windows` API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener`.

## Examples
The following methods create and close an offscreen document. Only a single Document can be open at a time. 

```js
chrome.offscreen.createDocument({
  url: chrome.runtime.getURL('off_screen.html'),

>   reasons: ['CLIPBOARD'],

Reviewer 2: It'd be nice to:

- Provide a list of each of the possible reasons, since developers probably won't know where to find those in the code (see [here](https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/api/offscreen/lifetime_enforcer_factories.cc;drc=503535015a7b373cc6185c69c991e01fda5da571;l=80?q=kReasonAndFactoryMethodPairs&ss=chromium)).
- Document that the browser may take steps to enforce the use (e.g requiring audio to play if audio playback is the justification).

I think a small tweak for each of these should be a blocker for merging because otherwise developers might simply copy this snippet and not update the reasons. We can probably just have a single line for point 2 for now though and have a follow-up to explicitly document how Chrome enforces each.

api: offscreen
---

## Manifest
You must declare the "offscreen" permission in the [extension manifest][1] to use the Offscreen API. For example:

```json
{
  "name": "My extension",
  ...
  "permissions": [
    "offscreen"
  ],
  ...
}
```
Pages loaded as offscreen documents are handled differently from other types of extensions pages. The extension's permissions carry over to offscreen documents, but extension API access is heavily limited. Currently, an offscreen document can only use the `chrome.runtime` APIs to send and receive messages; all other extension APIs are not exposed. Other notable differences between offscreen documents and normal pages are as follows:

* Offscreen document's URL must be a static HTML file bundled with the extension.
* Offscreen documents cannot be focused.

> * Offscreen documents cannot have their `opener` property set using the [`chrome.windows` API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener`.

Reviewer 3: ```suggestion
* Offscreen documents cannot have their `opener` property set using the [`chrome.windows` API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener()`.
```

## Reasons

> Reasons, listed [below](/docs/extensions/reference/offscreen/#type-Reason), are set upon  document creation to determine the document's lifespan. Currently, the AUDIO_PLAYBACK reason has specialized lifetime enforcement (CreateAudioLifetimeEnforcer). All others use generic idle detection   (CreateEmptyEnforcer).

Reviewer 2: Nit: Thanks for adding this, but can we link to https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/api/offscreen/audio_lifetime_enforcer.h and https://source.chromium.org/chromium/chromium/src/+/main:extensions/browser/api/offscreen/lifetime_enforcer_factories.cc. That at least gives an indication of the code we're referencing and a place to read more :)

Author: @jpmedley I haven't seen direct links to chromium source yet in other docs, let me know if this is within good practice.

Reviewer 3: Linking to source code is an anti-pattern on any platform. 
* It avoids using it as a crutch for things that aren't documented, but should be. 
* Avoids providing information in a programming language that the audience may not be familiar with. 

Please do a follow-up PR to add these to our documents. As we progress over the next year, we should have less and less reason to do patches like this.

Reviewer 2: The thing that worried me was that mentioning `CreateEmptyEnforcer` is completely useless to developers without telling them what it refers to. But happy for this to be a follow-up 👍

Reviewer 3: It should be a follow up.

> Reasons, listed [below](/docs/extensions/reference/offscreen/#type-Reason), are set upon  document creation to determine the document's lifespan. Currently, the AUDIO_PLAYBACK reason has specialized lifetime enforcement (CreateAudioLifetimeEnforcer). All others use generic idle detection   (CreateEmptyEnforcer).

Reviewer 3: ```suggestion
Reasons, listed [below](/docs/extensions/reference/offscreen/#type-Reason), are set upon document creation to determine the document's lifespan. Currently, the `AUDIO_PLAYBACK` reason has specialized lifetime enforcement (`CreateAudioLifetimeEnforcer`). All others use generic idle detection (`CreateEmptyEnforcer`).
```

## Examples

> The following methods create and close an offscreen document. Only a single Document can be open at a time. 

Reviewer 3: >Only a single Document can be open at a time.

This seems important enough that it should be in the description of an example. Please find a home for it above.

api: offscreen
---

## Manifest

> You must declare the "offscreen" permission in the [extension manifest][1] to use the Offscreen API. For example:

Reviewer 3: ```suggestion
You must declare the `"offscreen"` permission in the [extension manifest][1] to use the Offscreen API. For example:
```

```json
{
  "name": "My extension",
  ...
  "permissions": [
    "offscreen"
  ],
  ...
}
```

> Pages loaded as offscreen documents are handled differently from other types of extensions pages. The extension's permissions carry over to offscreen documents, but extension API access is heavily limited. Currently, an offscreen document can only use the `chrome.runtime` APIs to send and receive messages; all other extension APIs are not exposed. Other notable differences between offscreen documents and normal pages are as follows:

Reviewer 3: ```suggestion
Pages loaded as offscreen documents are handled differently from other types of extension pages. The extension's permissions carry over to offscreen documents, but extension API access is heavily limited. Currently, an offscreen document can only use the `chrome.runtime` APIs to send and receive messages; all other extension APIs are not exposed. Other notable differences between offscreen documents and normal pages are as follows:
```

> * Offscreen document's URL must be a static HTML file bundled with the extension.

Reviewer 3: ```suggestion
* An offscreen document's URL must be a static HTML file bundled with the extension.
```

* Offscreen documents cannot be focused.
* Offscreen documents cannot have their `opener` property set using the [`chrome.windows` API](docs/extensions/reference/windows/) method `windows.setSelfAsOpener()`.

> * An extension can only have one offscreen document open at a time. If the extension is running in split mode with an active incognito profile, both the normal and incognito profiles can each have one offscreen document. 

Reviewer 3: Question for a possible follow up PR: 'one offscreen document open at a time'? Does this also mean that I could have multiple offscreen documents for different purposes other than split mode as long as I only had one open at a time?
